### ISSUE TITLE
Inconsistency in Gradients Calculation of reduce_max

### ISSUE THREAD

**Issue by River861** [tensorflow/tensorflow#51681]
Hi, I found that the gradients calculation of `reduce_max` in tensorflow is different from other deep learning libraries such as theano, CNTK. And I want to know which algorithm is correct?

Here is an example code using TensorFlow2.6.0:
```python
import numpy as np
import tensorflow as tf

with tf.GradientTape() as tape:
    x = tf.Variable([[[0.6],
                      [0.6],
                      [0.3]]])

    y = tf.reduce_max(x)
g = tape.gradient(y, x)


print("gradients of max: ", g.numpy())
```
The result is:
```
gradients of max:  [[[0.5]
                     [0.5]
                     [0. ]]]
```

And this is the code using CNTK2.7:
```python
import numpy as np
import cntk as C

x = C.input_variable(shape=(1, 3, 1), needs_gradient=True)
x_val = np.array([[[0.6],
                   [0.6],
                   [0.3]]])

y = C.reduce_max(x)
g = y.grad({x: x_val})


print("gradients of max: ", g)
``` 
The result is:
```
gradients of max:  [[[[1.]
                      [1.]
                      [0.]]]]
```
The inconsistency exists when there are multiple max elements.

Any replies will be appreciated.
Thanks.

**Comment by kkimdev** [comment]
I think it's also good to check PyTorch and JAX before making a decision here.  First of all, the gradient is mathematically not defined in this case (left-limit and right-limit is different, 0 and 1), and for gradient-descent optimizing purpose, I think having a stable gradient norm with respect to floating point round off error is not a bad choice.

**Comment by sun51** [comment]
TF result is consistent with other framework (e.g. Jax). you can find the definition of max grad here  https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L234
